### ISSUE TITLE
Can FEMU simulate the difference in sequential and random read/write performance on SSD?

### ISSUE THREAD

**Issue by Borelset** [MoatLab/FEMU#137]
As the title, are the differences between sequential and random performance considered in the delay model?

**Comment by huaicheng** [comment]
The difference in seq/random write workloads stems mainly from FTL mgmt/GC efficiency, from this perspective, yes, FEMU can simulate the above scenario. It's better for you to compose some workloads and observe the performance difference directly on FEMU.

If you're talking about low-level NAND Flash commands optimized for handling adjacent seq/rand accesses, no, those are not modeled by in FEMU's NAND timing model.

**Comment by huaicheng** [comment]
Yes, though it comes from the parallelism model rather than from a seq/random knob.

bbssd (`femu_mode=1`) lays pages out across channels and LUNs and tracks a per-channel and per-LUN busy time. A sequential stream stripes across all of them and overlaps, while a random stream repeatedly lands on a LUN that is still busy with the previous operation and serialises against it. So the difference falls out of the geometry:

```
-device femu,...,femu_mode=1,nchs=8,luns_per_ch=8,pg_rd_lat=40000,pg_wr_lat=200000,blk_er_lat=2000000
```

More channels and LUNs widen the gap; `nchs=1,luns_per_ch=1` largely removes it.

Two other things shape it. Random *writes* also pay garbage collection: they scatter invalidations, so GC has more valid data to relocate, which shows up as write amplification (readable at byte 192 of `nvme smart-log -o binary`, with host and NAND page counts at 200/208/216). And `nand_cell_type` gives per-page-type latencies, so within a wordline some pages are slower than others.

`femu_mode=2` (NoSSD) has no timing model at all — every operation is instant — so use bbssd for this kind of study.
